src/utils/auth.py

```diff
@@ -0,0 +1,109 @@
+#!/usr/bin/env python3
+
+# Copyright 2024 Canonical Limited
+# See LICENSE file for licensing details.
+
+"""S3 connection utility classes and methods."""
+
+import secrets
+import string
+from dataclasses import dataclass
+
+from constants import (
```

Comment:
**suggestion** I would try to not use these imports here

Reply:
I removed the `AUTHENTICATION_DATABASE_NAME` as it will now be available as `database.dbname`, but these other constants are required. The `DEFAULT_ADMIN_USERNAME` is different from `database.username` because `DEFAULT_ADMIN_USERNAME` is the username of Kyuubi user while `database.username` is a PostgreSQL user.

Reply:
What about then to have this signature:

```
Authentication(database:DatabaseConnectionInfo, table_name: str)
```

This makes the class "pure", without using external, non-declared and implicit dependency. It should even make it easier to test

Reply:
This was required because we were hardcoding the admin function (see other comment) or similar functions. We can make a call if this can/should be ever be customized, but if we want to do that I would make the "DEFAULT_ADMIN_USERNAME" a class static attribute of `Authentication`. Again, following the functional tech talk I was giving, the pattern of injecting constants like this here is not very "pure". Hence the suggestion

Reply:
I agree to make this a class static attribute. I'll update this.